src/equations/lattice_boltzmann_3d.jl

```diff
@@ -0,0 +1,407 @@
+end
+
+
+"""
+    initial_condition_taylor_green_vortex(x, t, equations::LatticeBoltzmannEquations3D)
+
+Initialize the flow field to the Taylor-Green vortex setup
+"""
+function initial_condition_taylor_green_vortex(x, t, equations::LatticeBoltzmannEquations3D)
+  @unpack u0, rho0, L = equations
+
+  v1 =  u0 * sin(x[1] / L) * cos(x[2] / L) * cos(x[3] / L)
+  v2 = -u0 * cos(x[1] / L) * sin(x[2] / L) * cos(x[3] / L)
+  v3 = 0
+  p0 = (pressure(rho0, equations) +
+        rho0 * u0^2 / 16 * (cos(2 * x[1] / L) + cos(2 * x[2] / L)) * (cos(2 * x[3] / L) + 2))
+  rho = density(p0, equations)
+
+  return equilibrium_distribution(rho, v1, v2, v3, equations)
+end
+
+
+# Pre-defined source terms should be implemented as
+# function source_terms_WHATEVER(u, x, t, equations::LatticeBoltzmannEquations3D)
+
+
+# Calculate 1D flux in for a single point
+@inline function calcflux(u, orientation, equations::LatticeBoltzmannEquations3D)
+  if orientation == 1 # x-direction
+    v_alpha = equations.v_alpha1
+  elseif orientation == 2 # y-direction
+    v_alpha = equations.v_alpha2
+  else # z-direction
+    v_alpha = equations.v_alpha3
+  end
+  return v_alpha .* u
+end
+
+
+function flux_lax_friedrichs(u_ll, u_rr, orientation, equations::LatticeBoltzmannEquations3D)
+  if orientation == 1 # x-direction
+    v_alpha = equations.v_alpha1
+  elseif orientation == 2 # y-direction
+    v_alpha = equations.v_alpha2
+  else # z-direction
+    v_alpha = equations.v_alpha3
+  end
+  return 0.5 * ( v_alpha .* (u_ll + u_rr) - abs.(v_alpha) .* (u_rr - u_ll) )
+end
+
+
+"""
+    density(p::Real, equations::LatticeBoltzmannEquations3D)
+    density(u, equations::LatticeBoltzmannEquations3D)
+
+Calculate the macroscopic density from the pressure `p` or the particle distribution functions `u`.
+"""
+@inline density(p::Real, equations::LatticeBoltzmannEquations3D) = p / equations.c_s^2
+@inline density(u, equations::LatticeBoltzmannEquations3D) = sum(u)
+
+
+"""
+    velocity(u, orientation, equations::LatticeBoltzmannEquations3D)
+
+Calculate the macroscopic velocity for the given `orientation` (1 -> x, 2 -> y, 3 -> z) from the
+particle distribution functions `u`.
+"""
+@inline function velocity(u, orientation, equations::LatticeBoltzmannEquations3D)
+  if orientation == 1 # x-direction
+    v_alpha = equations.v_alpha1
+  elseif orientation == 2 # y-direction
+    v_alpha = equations.v_alpha2
+  else # z-direction
+    v_alpha = equations.v_alpha3
+  end
+
+  return sum(v_alpha .* u)/density(u, equations)
+end
+
+
+"""
+    velocity(u, equations::LatticeBoltzmannEquations3D)
+
+Calculate the macroscopic velocity vector from the particle distribution functions `u`.
+"""
+@inline function velocity(u, equations::LatticeBoltzmannEquations3D)
+  @unpack v_alpha1, v_alpha2, v_alpha3 = equations
+  rho = density(u, equations)
+
+  return SVector(sum(v_alpha1 .* u)/rho,
+                 sum(v_alpha2 .* u)/rho,
+                 sum(v_alpha3 .* u)/rho)
```

Comment:
```suggestion
  return SVector(dot(v_alpha1, u)/rho,
                 dot(v_alpha2, u)/rho,
                 dot(v_alpha3, u)/rho)
```
This improves the performance of `velocity` by ca. 15% on my system. We could get additional 5% on top by using `@avx` and writing out the loops, but that's probably not necessary right now.

Reply:
Wow, thanks for catching this! I agree, `@avx`ing is not necessary at the moment.